samples/nrf9160/serial_lte_modem/src/slm_at_host.c

```diff
@@ -354,35 +418,53 @@ int slm_at_host_init(void)
 	}
 
 	/* Choose which UART to use */
-	switch (uart_id) {
-	case UART_0:
+#if defined(CONFIG_SLM_CONNECT_UART_0)
 		uart_dev_name = SLM_UART_0_NAME;
-		break;
-	case UART_2:
+#elif defined(CONFIG_SLM_CONNECT_UART_2)
 		uart_dev_name = SLM_UART_2_NAME;
-		break;
-	default:
-		LOG_ERR("Unknown UART instance %d", uart_id);
+#else
+	LOG_ERR("Unsupported UART instance");
+	return -EINVAL;
+#endif
+	/* Initialize the UART module */
+	uart_dev = device_get_binding(uart_dev_name);
+	if (uart_dev == NULL) {
+		LOG_ERR("Cannot bind %s\n", uart_dev_name);
 		return -EINVAL;
 	}
-
-	err = at_notif_register_handler(NULL, response_handler);
-	if (err != 0) {
-		LOG_ERR("Can't register handler err=%d", err);
-		return err;
+	/* Wait for the UART line to become valid */
+	start_time = k_uptime_get_32();
+	do {
+		err = uart_err_check(uart_dev);
+		if (err) {
+			if (k_uptime_get_32() - start_time > K_MSEC(500)) {
+				LOG_ERR("UART check failed: %d. "
+					"UART initialization timed out.", err);
+				return -EIO;
+			}
+		}
+	} while (err);
+	/* Register async handling callback */
+	err = uart_callback_set(uart_dev, uart_callback, NULL);
+	if (err) {
+		LOG_ERR("Cannot set callback: %d", err);
+		return -EFAULT;
 	}
-
-	/* Initialize the UART module */
-	err = at_uart_init(uart_dev_name);
+	/* Power on UART module */
+	device_set_power_state(uart_dev, DEVICE_PM_ACTIVE_STATE,
+				NULL, NULL);
+	err = uart_rx_enable(uart_dev, uart_rx_buf[0], 1, K_FOREVER);
```

Comment:
What is the actual gain of switching to asynchronous API if responses are still received byte by byte? Is it only the question of being able to call `uart_tx` instead of `uart_poll_out` in a loop (what could be quite easily changed to use of TX IRQ)?

Reply:
This is due to the test result that, while using interrupt-driven Serial APIs, we cannot put the UART device into idle by disabling RX and powering off by Zephyr device power management. R&D suggested to use UARTE driver instead of Serial API, later @nordic-krch suggested to use Serial Async API, which was tested to be working.